Connection requests handling, refactoring

ercanucan opened the pull request:
@kormat @klausman: I updated the PR. It would be nice to get some feedback on this PR (at least for the Python code) when you get a chance.
@mwfarb: If you get a chance, it would be great to get some feedback from you on the front-end code (HTML, JavaScript).
This PR includes the following:
- Revised handling of join requests. Saves the keys, incoming Certificate, TRC
into the database and adds to config generation for deployment.
- Sending/Receiving of connection requests to/from coordination service.
- Ability to add the received connection reply into the active topology for deployment.
- Adapts existing code to the changes in SCION Coordination service API.
- Simplifies and tidies up Connection Requests front-end.
- Tidy up, refactoring and formatting changes.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/netsec-ethz/scion-web/52)
<!-- Reviewable:end -->

mwfarb commented:
@ercanucan I looked over the JS and HTML I have mostly general comments. The changes seem just fine. There are somethings to consider for lines that did not change, so minimal cleanup for readability.
- Needs license header: scion-web/ad_manager/static/js/ad_detail.js and topology_input.js.
- HTML under scion-web/ad_manager/templates/ad_manager has no license, but the overloading and inflation of partials may pollute the generated html with multiple license headers if all partial html files have them. I'd at least add a license to base.html and any files including JS <script> tags.
- Most html has mixed 2 and 4 spaced indents and should be consistent, probably 2. I'd keep the nested JS in the Html at 4.

kormat commented:
Reviewed 13 of 18 files at r1, 13 of 13 files at r2.
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks failed.

---

*[ad_manager/models.py, line 380 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Kd5vdNZn_M7D5zF8X_J:-Kd5vdNZn_M7D5zF8X_K:b-actbpc) ([raw file](https://github.com/netsec-ethz/scion-web/blob/ed5e32d37fb2b7995f64476b30bd01270d897b75/ad_manager/models.py#L380)):*
> ```Python
>     STATUS_OPTIONS = ['NONE', 'SENT', 'APPROVED', 'DECLINED']
>     LINK_TYPE = ['PARENT', 'CHILD', 'PEER', 'ROUTING']
>     OVERLAY_TYPE = ['IP', 'UDP/IP']
> ```

There should be 4 types here. `IPv4`, `IPv6`, `UDP/IPv4`, `UDP/IPv6`.

---

*[ad_manager/views.py, line 231 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Kd5wVZeqyRcsrqUhx04:-Kd5wVZeqyRcsrqUhx05:b-5wz9sk) ([raw file](https://github.com/netsec-ethz/scion-web/blob/ed5e32d37fb2b7995f64476b30bd01270d897b75/ad_manager/views.py#L231)):*
> ```Python
>         # TODO(ercanucan): create the ISD in DB if it's not yet in DB
>         AD.objects.update_or_create(
>             as_id=int(new_as[1]),
> ```

Why `int()`?

---

*[ad_manager/views.py, line 256 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Kd5xt-9bljCLRtRWNeI:-Kd5xt-9bljCLRtRWNeJ:b-t9b2yj) ([raw file](https://github.com/netsec-ethz/scion-web/blob/ed5e32d37fb2b7995f64476b30bd01270d897b75/ad_manager/views.py#L256)):*
> ```Python
>     Accepts or declines the join request. In case of accept, it assigns a new
>     AS ID to the requesting party and creates the certificate. This function
>     is only executed by a core AS.
> ```

This should be enforced re: core AS.

---

*[ad_manager/views.py, line 262 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Kd5yKAfTb1XtvO0FMhC:-Kd5yKAfTb1XtvO0FMhD:b-11dmw6) ([raw file](https://github.com/netsec-ethz/scion-web/blob/ed5e32d37fb2b7995f64476b30bd01270d897b75/ad_manager/views.py#L262)):*
> ```Python
>     own_isdas = ISD_AS(isd_as)
>     if status == REQ_APPROVED:
>         logger.info("new AS name = %s isd_as = %s", request.POST['newASname'],
> ```

What is an AS name?

---

*[ad_manager/views.py, line 354 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Kd6-LS8IkNXUwT4nQt7:-Kd6-LS8IkNXUwT4nQt8:bxas6jw) ([raw file](https://github.com/netsec-ethz/scion-web/blob/ed5e32d37fb2b7995f64476b30bd01270d897b75/ad_manager/views.py#L354)):*
> ```Python
>         created_by=request.user,
>         isd_to_join=isd_to_join,
>         join_as_a_core=(join_as_a_core == "True"),
> ```

It's usually best to handle case differences in situations like this. E.g. `join_as_a_core.lower() == "true"`

---

*[ad_manager/views.py, line 272 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4D2LxIjWHm_fuumCU:-Ke4D2LxIjWHm_fuumCV:bjz92h1) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L272)):*
> ```Python
>     else:
>         messages.info(request, 'Your join request with ID %s is declined '
>                                'by AS %s.' % (jr_id, join_reply['RespondIA']))
> ```

Can the user see the reason (if any) given when the request is declined?

---

*[ad_manager/views.py, line 311 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4DdDYWQvYq-C0YtLo:-Ke4DdDYWQvYq-C0YtLp:bjq2fz) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L311)):*
> ```Python
>         'RequesterId': request.POST['requester']
>     }
>     if status == REQ_APPROVED:
> ```

As this 21-line block of this is only relevant to approval, i suggest moving it into a separate method.

---

*[ad_manager/views.py, line 319 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4EBP8pIIk8xun_lOP:-Ke4EBP8pIIk8xun_lOQ:b-e09fml) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L319)):*
> ```Python
>         signing_as = AD.objects.get(as_id=own_isdas[1], isd=own_isdas[0])
>         signing_as_sig_priv_key = from_b64(signing_as.sig_priv_key)
>         signing_as_trc = str(signing_as.trc)
> ```

`signing_as.trc.to_json()`

---

*[ad_manager/views.py, line 331 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4Ekm03DHuA3FlxZTn:-Ke4Ekm03DHuA3FlxZTo:bj75k12) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L331)):*
> ```Python
>         join_rep_dict['RespondIA'] = str(own_isdas)
>         join_rep_dict['JoiningIACertificate'] = request_ia_chain.to_json()
>         join_rep_dict['RespondIACertificate'] = str(signing_as.certificate)
> ```

`.to_json()`

---

*[ad_manager/views.py, line 356 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4F-91Z597Gj7Qm5m7:-Ke4F-91Z597Gj7Qm5m8:b-5r3pph) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L356)):*
> ```Python
> @require_POST
> @login_required
> def request_new_as_id(request):
> ```

`request_join_isd` or similar is probably a better name.

---

*[ad_manager/views.py, line 452 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4L3VoXKVidRwwjFqJ:-Ke4L3VoXKVidRwwjFqK:b-yydfy9) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L452)):*
> ```Python
>                               UPLOAD_CONN_REQUEST_SVC, coord.account_id,
>                               coord.secret))
>         headers = {'contenttype': 'application/json'}
> ```

This block of code is repeated again and again, with very minor variations. It looks like it can probably be moved into a single utility function/method.

---

*[ad_manager/views.py, line 491 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4HPvzstdhPepwtSz0:-Ke4HPvzstdhPepwtSz1:b-2c3did) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L491)):*
> ```Python
>         con_req.overlay_type = data['overlay_type']
>         con_req.router_public_ip = data['router_info'].split(':')[0]
>         if con_req.overlay_type == ConnectionRequest.OVERLAY_TYPE[1]:  # UDP/IP
> ```

It would be cleaner to do `if "UDP" in con_req.overlay_type:`

---

*[ad_manager/views.py, line 502 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4HyDDeb44p8zbEavX:-Ke4HyDDeb44p8zbEavY:b8y6w6c) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L502)):*
> ```Python
>         return con_req
> 
>     def prepare_req_dic(self, con_req):
> ```

`prepare_req_dict` would be a bit better.

---

*[ad_manager/views.py, line 521 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4Ihj_s6Vq1_WlS6Zn:-Ke4Ihja8uzduoZNFdUi:b-5wz9sk) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L521)):*
> ```Python
>             "MTU": int(con_req.mtu),
>             "Bandwidth": int(con_req.bandwidth),
>             "Timestamp": iso_timestamp(int(time.time())),
> ```

Why `int()`?

---

*[ad_manager/views.py, line 594 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4K8qEJZuDDNGU7eRe:-Ke4K8qFtZCYqj2G4wQk:bhn4ivs) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L594)):*
> ```Python
>         overlay_type = data['accepted_overlay_type']
>         con_rep_dict["IP"] = router_info.split(':')[0]
>         if overlay_type == ConnectionRequest.OVERLAY_TYPE[1]:
> ```

Same comment re: checking if `UDP` is in the overlay_type`.

---

*[ad_manager/views.py, line 600 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4KEhDDaqVHAriKPKw:-Ke4KEhDDaqVHAriKPKx:bqwvt5l) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L600)):*
> ```Python
>         con_rep_dict["MTU"] = int(data['accepted_mtu'])
>         con_rep_dict["Bandwidth"] = int(data['accepted_bandwidth'])
>         con_rep_dict["Certificate"] = str(respond_as.certificate)
> ```

`.to_json()`?

---

*[ad_manager/views.py, line 1192 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4XUFIqjOdgP5qmfg_:-Ke4XUFIqjOdgP5qmfga:blr2qbl) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1192)):*
> ```Python
> 
> 
> def particular_topo_instance(tp, type_key):
> ```

This name is very unclear. What is it meant to do?

---

*[ad_manager/views.py, line 1206 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4RzqAntHzQlUQ1pld:-Ke4RzqAntHzQlUQ1ple:b1gk81o) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1206)):*
> ```Python
>                     'PortInternal')
>                 if type_key in ('BorderRouters', 'endhost'):
>                     continue  # Routers and endhost only know about external
> ```

What does this mean/do?

---

*[ad_manager/views.py, line 1222 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4SIeNuBd5Cl9Z-zfU:-Ke4SIeNuBd5Cl9Z-zfV:bo6vpb6) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1222)):*
> ```Python
>     """
>     routers = topo['BorderRouters']
>     complete_routers = {name: router for name, router in routers.items()
> ```

Dictionary comprehensions are usually better replaced with a simple for loop with an if statement.

---

*[ad_manager/views.py, line 1292 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4TYu14YDK958Fj70D:-Ke4TYu14YDK958Fj70E:bkv9pd) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1292)):*
> ```Python
>                 '"gen/ISD%s/AS%s/%s/zoo.cfg"')
>     cmd = cmd_tmpl % (isd_id, as_id, instance_name, class_path, instance_name,
>                       zoomain_env, isd_id, as_id, instance_name)
> ```

There's a few problems here:
- The first gen/ path is broken, as it doesn't have `ISD`, `AS` etc.
- The path creation is duplicated
- It's very hard to map the variables to the string interpolation.
- There's lots of quoting that isn't needed, making it hard to read.

A much better approach would be something like:
```
zk_path = "gen/ISD%s/AS%s/%s" % (isd_id, as_id, instance_name)
class_path = "%s:%s" % (zk_path, class_path)
cmd_parts = [
   "java", "-cp", class_path, "-Dzookeeper.log.filename=logs/%s.log" % instance_name,
    '"%s"' % zoomain_env, os.path.join(zk_path, "zoo.cfg")
]
return " ".join(cmd_parts)
```

---

*[ad_manager/views.py, line 1304 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4WnZwyR4TuXc9TA6u:-Ke4WnZwyR4TuXc9TA6v:bczm771) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1304)):*
> ```Python
>     config = configparser.ConfigParser()
>     env = 'PYTHONPATH=.,ZLOG_CFG="gen/dispatcher/dispatcher.zlog.conf"'
>     cmd = 'bash -c \'exec "bin/dispatcher" &>logs/dispatcher.OUT\''
> ```

If you use `"""` instead of `'`, you don't need escaping.
```
cmd = """bash -c 'exec bin/dispatcher &>logs/dispatcher.OUT'"""
```

---

*[ad_manager/views.py, line 1327 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4XMVzdcNnBz8o7KAg:-Ke4XMW-7SqVoPLk8oaw:b-1bct9f) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1327)):*
> ```Python
>     """
>     one_of_topology_path = os.path.join(instance_path, 'topology.yml')
>     one_of_topology = particular_topo_instance(tp, type_key)
> ```

I  don't understand the naming here. Why not just `path` and `topo`?

---

*[ad_manager/views.py, line 1411 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4_jeoA1XZF1yuGy3P:-Ke4_jepLmv5M6MN0sxb:b-6vmz2h) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1411)):*
> ```Python
>     cert_chain_path = get_cert_chain_file_path(instance_path,
>                                                ISD_AS.from_values(isd_id,
>                                                                   as_id),
> ```

Ahhh formatting :P Once a function call's arguments are too long to fit on two lines, move the first argument down to the next line, and flow them from there. It's much cleaner. In this case:
```
    cert_chain_path = get_cert_chain_file_path(
        instance_path, ISD_AS.from_values(isd_id, as_id), INITIAL_CERT_VERSION)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52)*
<!-- Sent from Reviewable.io -->

ercanucan commented:
Review status: 12 of 30 files reviewed at latest revision, 23 unresolved discussions.

---

*[ad_manager/models.py, line 380 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Kd5vdNZn_M7D5zF8X_J:-KeFZPbEChWxywkn2SVQ:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/ed5e32d37fb2b7995f64476b30bd01270d897b75/ad_manager/models.py#L380)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

There should be 4 types here. `IPv4`, `IPv6`, `UDP/IPv4`, `UDP/IPv6`.
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 231 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Kd5wVZeqyRcsrqUhx04:-KeIfw7a2dVM3t4_M4pb:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/ed5e32d37fb2b7995f64476b30bd01270d897b75/ad_manager/views.py#L231)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Why `int()`?
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 256 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Kd5xt-9bljCLRtRWNeI:-KeKBsR9iEWCwrfRWQBb:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/ed5e32d37fb2b7995f64476b30bd01270d897b75/ad_manager/views.py#L256)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This should be enforced re: core AS.
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 262 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Kd5yKAfTb1XtvO0FMhC:-KeOwupk2afKIzZ7Nz5U:b-fi4wyc) ([raw file](https://github.com/netsec-ethz/scion-web/blob/ed5e32d37fb2b7995f64476b30bd01270d897b75/ad_manager/views.py#L262)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

What is an AS name?
</blockquote></details>

Done. It is AS ID. Changed accordingly.

---

*[ad_manager/views.py, line 354 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Kd6-LS8IkNXUwT4nQt7:-KeJzxCw2LfAKtroCwRD:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/ed5e32d37fb2b7995f64476b30bd01270d897b75/ad_manager/views.py#L354)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

It's usually best to handle case differences in situations like this. E.g. `join_as_a_core.lower() == "true"`
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 272 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4D2LxIjWHm_fuumCU:-KeIg6KaRdOtIPnxh-_s:bpzlhtb) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L272)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Can the user see the reason (if any) given when the request is declined?
</blockquote></details>

scion-coord API supports this info field, but scion-web UI needs to provide these info windows to enter a reason while approving/declining join requests. I created a series of UI improvements and it's one of them (https://github.com/netsec-ethz/scion-web/issues/63). Will send this as a separate PR as this one has already become quite large.

---

*[ad_manager/views.py, line 311 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4DdDYWQvYq-C0YtLo:-KeKEkDKRRWgTz919cmT:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L311)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

As this 21-line block of this is only relevant to approval, i suggest moving it into a separate method.
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 319 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4EBP8pIIk8xun_lOP:-KeP1O05pjLSn9RPGSn7:b-w06ovh) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L319)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`signing_as.trc.to_json()`
</blockquote></details>

We persist the Certificate and TRCs as string in both scion-web and scion-coord. Therefore we send the stringified versions. 

---

*[ad_manager/views.py, line 331 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4Ekm03DHuA3FlxZTn:-KeP1OlKGd47ZNWQzcLe:br5tzug) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L331)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`.to_json()`
</blockquote></details>

Similar as above. We transmit them as strings. Removed the unnecessary string casting here. (`to_json()` function of `CertificateChain` already gives us a stringified json)

---

*[ad_manager/views.py, line 356 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4F-91Z597Gj7Qm5m7:-KeOxgqilKjkix_tFYEP:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L356)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`request_join_isd` or similar is probably a better name.
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 452 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4L3VoXKVidRwwjFqJ:-KeU9p07HU1743GoJERQ:b-uwf57e) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L452)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This block of code is repeated again and again, with very minor variations. It looks like it can probably be moved into a single utility function/method.
</blockquote></details>

Done. Separated it out to a function called `post_req_to_scion_coord` and put it inside `util.py`.

---

*[ad_manager/views.py, line 491 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4HPvzstdhPepwtSz0:-KeIfaOS5oXV_XUfc_yK:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L491)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

It would be cleaner to do `if "UDP" in con_req.overlay_type:`
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 502 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4HyDDeb44p8zbEavX:-KeIfbaxo6zkWmZVaWLx:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L502)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`prepare_req_dict` would be a bit better.
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 521 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4Ihj_s6Vq1_WlS6Zn:-KeK-3PqUy9lK--Vlyen:bfmtp1b) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L521)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Why `int()`?
</blockquote></details>

Done. Not casting to int anymore (was from the existing code)

---

*[ad_manager/views.py, line 594 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4K8qEJZuDDNGU7eRe:-KeK0YT9e14McpNdCy1h:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L594)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Same comment re: checking if `UDP` is in the overlay_type`.
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 600 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4KEhDDaqVHAriKPKw:-KeP1_ur-829WG_fpSDq:bbe8oy2) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L600)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

`.to_json()`?
</blockquote></details>

Same as above. (Removed the unnecessary string casting)

---

*[ad_manager/views.py, line 1192 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4XUFIqjOdgP5qmfg_:-KeUmu4viP7PI1xFzyxb:b-najt0u) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1192)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This name is very unclear. What is it meant to do?
</blockquote></details>

Changed the name to `topo_instance` (the name was a remainder from the existing implementation). This is the workaround function for NAT cases until we have the new topology format. It generates the topology for a given service type based on whether internal address/port are specified in the topology input. If there are internal addresses provided, then it generates the topology so that these internal addresses are used for those services to bind to. Border routers and end-host always uses the public addresses in the topology file as we can run them only in non-NATted environments.

---

*[ad_manager/views.py, line 1206 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4RzqAntHzQlUQ1pld:-KeUoMn0urfeD5lqZZdK:bqalmf4) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1206)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

What does this mean/do?
</blockquote></details>

explained above.

---

*[ad_manager/views.py, line 1222 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4SIeNuBd5Cl9Z-zfU:-KeUKuX60rKDvQpNmCpS:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1222)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Dictionary comprehensions are usually better replaced with a simple for loop with an if statement.
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 1292 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4TYu14YDK958Fj70D:-KeK8iw-gyfLG-j5_ppD:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1292)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

There's a few problems here:
- The first gen/ path is broken, as it doesn't have `ISD`, `AS` etc.
- The path creation is duplicated
- It's very hard to map the variables to the string interpolation.
- There's lots of quoting that isn't needed, making it hard to read.

A much better approach would be something like:
```
zk_path = "gen/ISD%s/AS%s/%s" % (isd_id, as_id, instance_name)
class_path = "%s:%s" % (zk_path, class_path)
cmd_parts = [
   "java", "-cp", class_path, "-Dzookeeper.log.filename=logs/%s.log" % instance_name,
    '"%s"' % zoomain_env, os.path.join(zk_path, "zoo.cfg")
]
return " ".join(cmd_parts)
```
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 1304 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4WnZwyR4TuXc9TA6u:-KeK190y-sDNL_Nwmsmi:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1304)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

If you use `"""` instead of `'`, you don't need escaping.
```
cmd = """bash -c 'exec bin/dispatcher &>logs/dispatcher.OUT'"""
```
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 1327 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4XMVzdcNnBz8o7KAg:-KeULQHnrvb9r8jtsy_G:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1327)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

I  don't understand the naming here. Why not just `path` and `topo`?
</blockquote></details>

Done.

---

*[ad_manager/views.py, line 1411 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4_jeoA1XZF1yuGy3P:-KeK0w6VRsbwJtFBR_qs:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L1411)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Ahhh formatting :P Once a function call's arguments are too long to fit on two lines, move the first argument down to the next line, and flow them from there. It's much cleaner. In this case:
```
    cert_chain_path = get_cert_chain_file_path(
        instance_path, ISD_AS.from_values(isd_id, as_id), INITIAL_CERT_VERSION)
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52)*
<!-- Sent from Reviewable.io -->

ercanucan commented:
@mwfarb addressed your comments: Added license headers to the files you mentioned. Changed the indentation of the HTML files to 2 spaces, kept the nested JS at 4 spaces.
@kormat addressed your comments.

---

Review status: 12 of 30 files reviewed at latest revision, 23 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-:-KeVNt5a1IjIIiogbCDw:bbvvt0n)*
<!-- Sent from Reviewable.io -->

mwfarb commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 12 of 30 files reviewed at latest revision, 23 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-:-KeYq6DmmrtlPD8vjjkT:bnfp4nl)*
<!-- Sent from Reviewable.io -->

kormat commented:
Reviewed 18 of 18 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[ad_manager/views.py, line 319 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4EBP8pIIk8xun_lOP:-KeZA3wDTJEvAzDBAepG:b-f60krr) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L319)):*
<details><summary><i>Previously, ercanucan (ercanucan) wrote…</i></summary><blockquote>

We persist the Certificate and TRCs as string in both scion-web and scion-coord. Therefore we send the stringified versions. 
</blockquote></details>

This isn't about the type, `to_json()` still produces a string. `__str__` is designed, in general, to produce human-readable output. Therefore it may contain extra stuff that is not JSON, and thus may break later parsing. Using a dedicated method to produce exactly (and only) what you want is preferred because of this.

---

*[ad_manager/views.py, line 331 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4Ekm03DHuA3FlxZTn:-KeZAYlJo-R3M5dvemSa:bo4btz7) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L331)):*
<details><summary><i>Previously, ercanucan (ercanucan) wrote…</i></summary><blockquote>

Similar as above. We transmit them as strings. Removed the unnecessary string casting here. (`to_json()` function of `CertificateChain` already gives us a stringified json)
</blockquote></details>

Addressed above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52)*
<!-- Sent from Reviewable.io -->

ercanucan commented:
Review status: 28 of 30 files reviewed at latest revision, 1 unresolved discussion.

---

*[ad_manager/views.py, line 319 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4EBP8pIIk8xun_lOP:-KedOaIys815uYD80Dsn:bhr72z) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L319)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This isn't about the type, `to_json()` still produces a string. `__str__` is designed, in general, to produce human-readable output. Therefore it may contain extra stuff that is not JSON, and thus may break later parsing. Using a dedicated method to produce exactly (and only) what you want is preferred because of this.
</blockquote></details>

Done. Yes, I see what you mean. Now sending all the TRCs and Certs after calling `to_json()` on them. Please note that `to_json()` function of TRC needs the fix from https://github.com/netsec-ethz/scion/pull/998.

---

*[ad_manager/views.py, line 331 at r2](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-Ke4Ekm03DHuA3FlxZTn:-KedP-e3IGMpZiCSH-SA:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/c644b7c677fdd3e07a88fa99273942a80e911583/ad_manager/views.py#L331)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Addressed above.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52)*
<!-- Sent from Reviewable.io -->

kormat commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/52#-:-KehBFMWZ-fbC5z3s9sE:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ercanucan commented:
Thanks! Merging.
